components/server/src/auth/verification-service.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { User } from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable, postConstruct } from "inversify";
+import { Config } from "../config";
+import { Twilio } from "twilio";
+import { ServiceContext } from "twilio/lib/rest/verify/v2/service";
+import { UserDB, WorkspaceDB } from "@gitpod/gitpod-db/lib";
+
+@injectable()
+export class VerificationService {
+    @inject(Config) protected config: Config;
+    @inject(WorkspaceDB) protected workspaceDB: WorkspaceDB;
+    @inject(UserDB) protected userDB: UserDB;
+
+    protected verifyService: ServiceContext;
+
+    @postConstruct()
+    protected initialize(): void {
+        if (this.config.twilioConfig) {
+            const client = new Twilio(this.config.twilioConfig.accountSID, this.config.twilioConfig.authToken);
+            this.verifyService = client.verify.v2.services(this.config.twilioConfig.serviceID);
+        }
+    }
+
+    public async needsVerification(user: User): Promise<boolean> {
+        if (!this.config.twilioConfig) {
+            return false;
+        }
+        // if (!!user.additionalData?.lastVerificationTime) {
+        //     return false;
+        // }
+        // if (await this.isTrusted(user)) {
+        //     await this.markVerified(user);
+        //     return false;
+        // }
+        return true;
+    }
+
+    public async markVerified(user: User): Promise<User> {
+        if (!user.additionalData) {
+            user.additionalData = {};
+        }
+        user.additionalData!.lastVerificationTime = new Date().toISOString();
+        await this.userDB.updateUserPartial(user);
+        return user;
+    }
+
+    // private async isTrusted(user: User): Promise<boolean> {
+    //     // we only want to ask new users, so if a user has ever started workspaces before, we trust them.
+    //     // there might be a few false positives because all workspaces might have been deleted.
+    //     const workspaces = await this.workspaceDB.findWorkspacesByUser(user.id);
+    //     return workspaces.length !== 0;
```

Comment:
@geropl I think it makes sense to check for usage instead?

Reply:
Jep, that could work as an indicator, with a threshold > what we see from abusers.